controllers/accounts.go

```diff
@@ -256,45 +324,227 @@ func (ac *AccountsController) SetupPasswordFinalize(w http.ResponseWriter, r *ht
 	if err != nil {
 		util.RenderErrorResponse(w, r, http.StatusInternalServerError, err)
 		return
 	}
 
-	if err := ac.ds.DeleteVerification(verification.ID); err != nil {
+	if twoFADetails.TOTP {
+		util.RenderErrorResponse(w, r, http.StatusBadRequest, util.ErrTOTPAlreadyEnabled)
+		return
+	}
+
+	key, err := ac.twoFAService.GenerateAndStoreTOTPKey(session.AccountID, session.Email)
+	if err != nil {
 		util.RenderErrorResponse(w, r, http.StatusInternalServerError, err)
 		return
 	}
 
+	response := TwoFAInitResponse{
+		URL: key.URL(),
+	}
+
+	if requestData.GenerateQR {
+		qrCode, err := ac.twoFAService.GenerateTOTPQRCode(key)
+		if err != nil {
+			util.RenderErrorResponse(w, r, http.StatusInternalServerError, err)
+			return
+		}
+		response.QRCode = &qrCode
+	}
+
 	render.Status(r, http.StatusOK)
-	render.JSON(w, r, &PasswordFinalizeResponse{
-		AuthToken: authToken,
-	})
+	render.JSON(w, r, response)
+}
+
+// @Summary Finalize TOTP 2FA setup
+// @Description Complete the TOTP 2FA setup process by validating a TOTP code and enabling 2FA for the account.
+// @Tags Accounts
+// @Accept json
+// @Produce json
+// @Param Authorization header string true "Bearer + auth token"
+// @Param Brave-Key header string false "Brave services key (if one is configured)"
+// @Param request body TwoFAFinalizeRequest true "2FA finalization request"
+// @Success 200 {object} TwoFAFinalizeResponse
+// @Failure 400 {object} util.ErrorResponse
+// @Failure 401 {object} util.ErrorResponse
+// @Failure 500 {object} util.ErrorResponse
+// @Router /v2/accounts/2fa/totp/finalize [post]
+func (ac *AccountsController) SetupTOTPFinalize(w http.ResponseWriter, r *http.Request) {
+	session := r.Context().Value(middleware.ContextSession).(*datastore.SessionWithAccountInfo)
+
+	var requestData TwoFAFinalizeRequest
+	if !util.DecodeJSONAndValidate(w, r, &requestData) {
+		return
+	}
+
+	if err := ac.twoFAService.ValidateTOTPCode(session.AccountID, requestData.Code); err != nil {
+		if errors.Is(err, util.ErrBadTOTPCode) {
+			util.RenderErrorResponse(w, r, http.StatusBadRequest, err)
+		} else {
+			util.RenderErrorResponse(w, r, http.StatusInternalServerError, err)
+		}
+		return
+	}
+
+	if err := ac.ds.SetTOTPSetting(session.AccountID, true); err != nil {
+		util.RenderErrorResponse(w, r, http.StatusInternalServerError, err)
+		return
+	}
+
+	// Check if recovery key already exists
+	hasRecoveryKey, err := ac.ds.HasRecoveryKey(session.AccountID)
+	if err != nil {
+		util.RenderErrorResponse(w, r, http.StatusInternalServerError, err)
+		return
+	}
+
+	var response TwoFAFinalizeResponse
+
+	// Only generate a recovery key if one doesn't exist
+	if !hasRecoveryKey {
+		// Generate and store a recovery key
+		recoveryKey, err := ac.twoFAService.GenerateAndStoreRecoveryKey(session.AccountID)
+		if err != nil {
+			util.RenderErrorResponse(w, r, http.StatusInternalServerError, err)
+			return
+		}
+
+		response.RecoveryKey = &recoveryKey
+		response.TwoFADisabled = true
```

Comment:
I think that's wrong. We just generated a new recovery key because one didn't already exist (i.e. user is probably enabling 2FA for the first time).

As I understand it from the comments above, this flag is for telling the client that a recovery code was just used to disable 2FA.
```suggestion
```